### ISSUE TITLE
Numerical Instabilities and other non-reproducible behaviour

### ISSUE THREAD

**Issue by sheepforce** [pyscf/pyscf#1123]
Dear developers,
I am maintaining the [PySCF package in the NixOS distribution and the nixpkgs collection](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/pyscf/default.nix). We have recently upgraded to the 2.0.1 release. During the [update PR in nixpkgs](https://github.com/NixOS/nixpkgs/pull/144253) my reviewers noticed, that the test suite fails for them with heavy numerical discrepancies, while on all my machines everything builds fine. Please note that Nixpkgs ensures strict sandboxing and reproducibility of build environments down to the exact hashes of all dependencies. Therefore we can exclude different build environments and settings. PySCF passes the tests on all of my machines (Intel Xeons or i7s), on one part of our build infrastructure (called ofBorg, also Intel), but it fails on the machines of my reviewers (@markuskowa), and also on other parts of our build infrastructure (called Hydra, uses AMD Epyc instances).

We have noticed numerically instable behaviour with OpenBLAS previously (https://github.com/markuskowa/NixOS-QChem/issues/52 and https://github.com/NixOS/nixpkgs/issues/124250), but those were related to AVX512 kernels and the problem hasn't occured since then.

Here are some of the errors the PySCF test suite _may_ throw:

- https://github.com/NixOS/nixpkgs/pull/144253#pullrequestreview-803917750 (too many open files without any reason)
- https://github.com/NixOS/nixpkgs/pull/144253#issuecomment-977846458 (unreasonable MP2 memory issues)
- https://hydra.nixos.org/build/159904468/nixlog/1 (too many open files again)
- https://hydra.nixos.org/build/159803621/nixlog/2 (AArch64, numerically heavily incorrect)

Building with MKL seems to solve the issues (https://github.com/NixOS/nixpkgs/pull/144253#issuecomment-985465196), but still fails in some tests with minor numerical issues.

Therefore, we have a lot of non-reproducible and worrying behaviour, albeit we can ensure exactly reproducible build environments with the excpetion of hardware, of course. I have a hard time tracking these issues down. Are you aware of any strange behaviour of PySCF with specific BLAS implementations? Psi4 e.g. has problems with threaded OpenBLAS.

**Comment by jamesETsmith** [comment]
@sheepforce and @markuskowa, first off, thanks for all your help maintaining the NixOS distribution of the PySCF! We really appreciate your work making PySCF accessible to more users!

I haven't come across any problems with OpenBLAS in the past (even on AMD machines), but those were versions of PySCF < 2.0.1. I have access to some AMD Epyc machines and will try out building PySCF with OpenBLAS on them and get back to you.

**Comment by jamesETsmith** [comment]
@sheepforce, I haven't been able to reproduce the numerical errors and on my system only a single test (`test_avas`) fails. I build PySCF (5bfe7d1) with gcc@10.2.0, openblas@0.3.15, and I'm using python@3.9.7. For reference my processor is an AMD EPYC 7742.

I don't think we should see problems with the threading in the BLAS implementation when because PySCF switches it off here (since BLAS calls are usually inside parallelized regions):

https://github.com/pyscf/pyscf/blob/5bfe7d1f96f9705c3b69a6da1ff148924e15ae4c/pyscf/__init__.py#L44-L47

The tests are also run serially (see the script used in CI below)

https://github.com/pyscf/pyscf/blob/5bfe7d1f96f9705c3b69a6da1ff148924e15ae4c/.github/workflows/run_tests.sh#L2

## Suggestions
The first (bad) idea I have for the "too many open files" problem would be to try and increase the maximum allowed (with `ulimit`). This is definitely not a satisfactory solution long term, but may help things run on your system properly for now.

After a bit more searching, this file limit problem shows up when running the ADC module tests and it looks like that module doesn't run `os.close(file_descriptor)` for tempfiles which can leave them open.

See the ADC code here (the file_descriptor itself isn't saved):
https://github.com/pyscf/pyscf/blob/1de8c145abb3e1a7392df9118e8062e6fe6bde00/pyscf/adc/uadc_ao2mo.py#L228-L232

And the discussion about `tempfile` usage here: https://bugs.python.org/issue42830.

(Full disclosure, I'm not very familiar with the ADC module, so I may be missing something here).

**Comment by markuskowa** [comment]
> The first (bad) idea I have for the "too many open files" problem would be to try and increase the maximum allowed (with ulimit). This is definitely not a satisfactory solution long term, but may help things run on your system properly for now.

The problem appears in nix's central CI (e.g. https://hydra.nixos.org/build/159904468/nixlog/1), where can not manipulate ulimit.

**Comment by sunqm** [comment]
Too many opened files is a known issue when running tests with nose. A hook function `tearDownModule` is therefore added in some test modules to close files. But this hook function was not implemented still in many test modules. Also tearDownModule function may not be triggered if some errors were raised in any tests of a test module. If `ulimit -n` was set to a relatively small number like 500, it's very possible to get error of "too many opened files". 

I think it's safe to increase `ulimit -n` to remove the "too many opened files" error if applicable in your system.

**Comment by sheepforce** [comment]
As @markuskowa said, we can not increase ulimit -n in the build sandbox and on the CI server in general. The only option currently seems therefore to disable all ADC tests. I guess that is an okayish solution for packaging right now.

Then there is the open point of the numerical issues. They are large and there are many on AArch64 (https://hydra.nixos.org/build/159803621/nixlog/2). Any idea what to do about those? :slightly_smiling_face:

**Comment by sunqm** [comment]
Is quadmath available in your system? quadmath is used to generate two-electron integrals. It may be another source of numerical instabilities

**Comment by sheepforce** [comment]
No, quadmath is not part of the package expression and not in our dependency graph.

**Comment by sunqm** [comment]
Yes, that's what I suspect. Without quadmath some 2-electron integrals may have numerical uncertainty.

**Comment by verena-neufeld** [comment]
Hi @sheepforce , thanks for the work and posting this issue here. Has it been resolved now?

**Comment by sheepforce** [comment]
I'm afraid I can't give a completely definitive answer. We've disabled PySCF support on all non x86_64 platforms as the tests always kept failing on e.g. aarch64. However, we haven't checked for a long time in automatic builds on aarch64-linux and aarch64-darwin and I don't have access to such machines. There is also a short list of tests I've had to disable on all platforms, as they are always numerically slightly off (see https://github.com/NixOS/nixpkgs/blob/9b19f5e77dd906cb52dade0b7bd280339d2a1f3d/pkgs/development/python-modules/pyscf/default.nix#L61C9-L61C9). 
At least on x86-64-linux the builds mostly look fine over the last months: 
  
  * https://hydra.nixos.org/job/nixpkgs/trunk/python311Packages.pyscf.x86_64-linux
  * https://nix-qchem-hydra.fysik.su.se//job/NixOS-QChem/master-unstable/qchem.tests.pyscf

 (the failures are mostly caused by larger upgrades in the python ecosystem). So at least on x86-64 it works stable now, yes. I can't tell for aarch64 where we observed most problems.

**Comment by kvmto** [comment]
> I'm afraid I can't give a completely definitive answer. We've disabled PySCF support on all non x86_64 platforms as the tests always kept failing on e.g. aarch64. However, we haven't checked for a long time in automatic builds on aarch64-linux and aarch64-darwin and I don't have access to such machines. There is also a short list of tests I've had to disable on all platforms, as they are always numerically slightly off (see https://github.com/NixOS/nixpkgs/blob/9b19f5e77dd906cb52dade0b7bd280339d2a1f3d/pkgs/development/python-modules/pyscf/default.nix#L61C9-L61C9). At least on x86-64-linux the builds mostly look fine over the last months:
> 
> * https://hydra.nixos.org/job/nixpkgs/trunk/python311Packages.pyscf.x86_64-linux
> * https://nix-qchem-hydra.fysik.su.se//job/NixOS-QChem/master-unstable/qchem.tests.pyscf
> 
> (the failures are mostly caused by larger upgrades in the python ecosystem). So at least on x86-64 it works stable now, yes. I can't tell for aarch64 where we observed most problems.
> 
> EDIT:
> 
> > Without quadmath some 2-electron integrals may have numerical uncertainty.
> 
> Meanwhile I've checked this point: That was a mistake on my side 😬 Our gcc supports quadmath, of course.

Dear @sheepforce,

Thank you for your previous response. I have a question regarding repeatability across different executions on the same machine. In my project, I am using PySCF to calculate single-body and two-body integrals. I've noticed that some level of repeatability can be achieved when setting OMP_NUM_THREADS=1. However, unfortunately, there are fluctuations in the last decimal places of the results, which significantly affect the accuracy of my application.

Do you have any solutions or suggestions on how to address this issue?:)

**Comment by sheepforce** [comment]
> Do you have any solutions or suggestions on how to address this issue?:)

Oh I wish, but unfortunately not. 😬 I sometimes even see differences between AMD and Intel CPUs, also with other codes and even when forcing BLAS to use the same kernels. Floating point arithmetics is subtle, I suppose 😅
